Assets/Scripts/StandardSamples/Services/PlayerDataStorageService.cs

```diff
@@ -0,0 +1,468 @@
+                    {
+                        downloadTask.ResultCode = data.ResultCode;
+                        tcs.SetException(new Exception($"{nameof(PlayerDataStorageService)} ({nameof(DownloadFile)}): Failed to download file. Result code {data.ResultCode}"));
+                        _playerDataStorageSemaphore.Release();
+                        return;
+                    }
+
+                    var copyFileMetadataOptions = new CopyFileMetadataByFilenameOptions { Filename = fileName, LocalUserId = userId };
+                    Result metadataRetrievalResult = EOSManager.Instance.GetPlayerDataStorageInterface().CopyFileMetadataByFilename(ref copyFileMetadataOptions, out FileMetadata? fileMetadata);
+
+                    if (metadataRetrievalResult != Result.Success)
+                    {
+                        downloadTask.ResultCode = metadataRetrievalResult;
+                        tcs.SetException(new Exception($"{nameof(PlayerDataStorageService)} ({nameof(DownloadFile)}): Failed to retrieve file metadata. Result code {metadataRetrievalResult}"));
+                        _playerDataStorageSemaphore.Release();
+                        return;
+                    }
+
+                    // FileMetaData.FileSizeBytes includes header information, which doesn't need to be considered while downloading
+                    downloadTask.SetDataBufferSize((uint)fileMetadata?.UnencryptedDataSizeBytes);
+
+                    // The returned EOSPlayerDataStorageTransferTask holds the callbacks for file progress and finishing
+                    ReadFileOptions readOptions = new ReadFileOptions
+                    {
+                        LocalUserId = userId,
+                        Filename = fileName,
+                        ReadChunkLengthBytes = MAX_CHUNK_SIZE,
+                        ReadFileDataCallback = downloadTask.OnFileDataReceived,
+                        FileTransferProgressCallback = downloadTask.OnFileTransferProgressUpdated
+                    };
+
+                    downloadTask.TransferRequest = EOSManager.Instance.GetPlayerDataStorageInterface().ReadFile(ref readOptions, null,
+                        (ref ReadFileCallbackInfo info) => 
+                    {
+                        downloadTask.OnFileReceived(ref info);
+                        _playerDataStorageSemaphore.Release();
+                    });
+
+                    // If no valid handle is returned, then error out
+                    if (downloadTask.TransferRequest == null)
+                    {
+                        downloadTask.ResultCode = Result.UnexpectedError;
+                        tcs.SetException(new Exception($"{nameof(PlayerDataStorageService)} ({nameof(DownloadFile)}): Bad handle returned for reading file, can not download file."));
+                        _playerDataStorageSemaphore.Release();
+                        return;
+                    }
+                });
```

Comment:
This code is doing three discrete things: 

1. Querying the file
2. Copying the file meta data
3. Reading the file

I think that these three tasks should be split into individual methods. This will make the code a lot easier to read and consequently to debug.